compiler/aarch64/codegen/BinaryEvaluator.cpp

```diff
@@ -1153,16 +1299,153 @@ logicBinaryEvaluator(TR::Node *node, TR::InstOpCode::Mnemonic regOp, TR::InstOpC
    return trgReg;
    }
 
+/**
+ * @brief Generates ubfm instruction for shift and mask operation if possible
+ *
+ * @param[in] andNode: node with and opcode
+ * @param[in]      cg: code generator
+ *
+ * @returns register which contains the result of the operation. NULL if the operation cannot be encoded in ubfm instruction.
+ */
+static TR::Register *
+generateUBFMForShiftAndMask(TR::Node *andNode, TR::CodeGenerator *cg)
+   {
+   TR::Node *shiftNode = NULL;
+   TR::Node *maskNode = NULL;
+   if (andNode->getFirstChild()->getOpCode().isShift())
+      {
+      shiftNode = andNode->getFirstChild();
+      maskNode = andNode->getSecondChild();
+      }
+   else if (andNode->getSecondChild()->getOpCode().isShift())
+      {
+      shiftNode = andNode->getSecondChild();
+      maskNode = andNode->getFirstChild();
+      }
+   if (shiftNode == NULL)
+      {
+      return NULL;
+      }
+
+   if ((shiftNode->getReferenceCount() == 1) &&
+       (shiftNode->getRegister() == NULL) &&
+       shiftNode->getSecondChild()->getOpCode().isLoadConst() &&
+       maskNode->getOpCode().isLoadConst())
+      {
+      const bool is64bit = shiftNode->getDataType().isInt64();
+      int64_t shiftValue = shiftNode->getSecondChild()->getConstValue();
+      if ((shiftValue <= 0) || (shiftValue > (is64bit ? 63 : 31)))
+         {
+         return NULL;
+         }
+      uint64_t maskValue = is64bit ? maskNode->getLongInt() : static_cast<uint32_t>(maskNode->getInt());
+
+      if (shiftNode->getOpCode().isLeftShift())
+         {
+         if (maskValue < (static_cast<uint64_t>(1) << shiftValue))
+            {
+            /*
+             * The result is always 0 in this case.
+             */
+            TR::Register *reg = cg->allocateRegister();
+            generateTrg1ImmInstruction(cg, TR::InstOpCode::movzx, andNode, reg, 0);
+            andNode->setRegister(reg);
+            cg->recursivelyDecReferenceCount(shiftNode);
+            cg->decReferenceCount(maskNode);
+            return reg;
+            }
+
+         uint64_t shiftedMask = (maskValue >> shiftValue);
+         if (((shiftedMask & 1) == 1) && contiguousBits(shiftedMask))
+            {
+            /*
+             * shiftedMask has consecutive 1s from the least significant bits.
+             * So, this shift and mask operation is copying consecutive bits starting from the least significant bits of the source register
+             * to bit position shiftValue of the destination register.
+             */
+            uint32_t width = populationCount(shiftedMask);
+            uint32_t imms = (width - 1);
+            uint32_t immr = (is64bit ? 64 : 32) - shiftValue;
+
+            TR::Register *reg = cg->allocateRegister();
```

Comment:
The same comment for the reference count of `sourceNode`.

Reply:
force-pushed the updated commit, too.